src/main/java/com/spotify/annoy/ANNIndex.java

```diff
@@ -69,43 +71,72 @@ public ANNIndex(final int dimension,
 
   private void load(final String filename) throws IOException {
     memoryMappedFile = new RandomAccessFile(filename, "r");
-    int fileSize = (int) memoryMappedFile.length();
-    // We only support indexes <4GB as a result of ByteBuffer using an int index
-    annBuf = memoryMappedFile.getChannel().map(
-            FileChannel.MapMode.READ_ONLY, 0, fileSize);
-    annBuf.order(ByteOrder.LITTLE_ENDIAN);
+    long fileSize = memoryMappedFile.length();
+    int buffIndex = (int) Math.ceil(((double) fileSize) / BLOCK_SIZE) - 1;
+    int rest = (int) (fileSize % BLOCK_SIZE);
+    int blockSize = (rest > 0 ? rest : BLOCK_SIZE);
+    long position = Math.max(0, fileSize - blockSize);
+
+    boolean process = true;
     int m = -1;
-    for (int i = fileSize - NODE_SIZE; i >= 0; i -= NODE_SIZE) {
-      int k = annBuf.getInt(i);  // node[i].n_descendants
-      if (m == -1 || k == m) {
-        roots.add(i);
-        m = k;
-      } else {
-        break;
+    long index = fileSize;
+    while(blockSize > 0) {
+      MappedByteBuffer annBuf = memoryMappedFile.getChannel().map(
+              FileChannel.MapMode.READ_ONLY, position, blockSize);
+      annBuf.order(ByteOrder.LITTLE_ENDIAN);
+
+      buffers.put(buffIndex--, annBuf);
+
+      for (int i = blockSize - NODE_SIZE; process && i >= 0; i -= NODE_SIZE) {
+        index -= NODE_SIZE;
+        int k = annBuf.getInt(i);  // node[i].n_descendants
+        if (m == -1 || k == m) {
+          roots.add(index);
+          m = k;
+        } else {
+          process = false;
+          break;
+        }
       }
+      blockSize = (int) Math.min(BLOCK_SIZE, position);
+      position = position - blockSize;
```

Comment:
??? `blockSize` is always `BLOCK_SIZE` here, except for the last block which we did first, and position should decrement by blockSize first, not second.